src/core/overlay/position/connected-position-strategy.ts

```diff
@@ -0,0 +1,204 @@
+  }
+
+
+  /**
+   * Updates the position of the overlay element, using whichever preferred position relative
+   * to the origin fits on-screen.
+   */
+  apply(element: HTMLElement): Promise<void> {
+    // We need the bounding rects for the origin and the overlay to determine how to position
+    // the overlay relative to the origin.
+    const originRect = this._origin.getBoundingClientRect();
+    const overlayRect = element.getBoundingClientRect();
+
+    // We use the viewport rect to determine whether a position would go off-screen.
+    const viewportRect = this._viewportRuler.getViewportRect();
+    let firstOverlayPoint: Point = null;
+
+    // We want to place the overlay in the first of the preferred positions such that the
+    // overlay fits on-screen.
+    for (let pos of this._preferredPositions) {
+      // Get the (x, y) point of connection on the origin, and then use that to get the
+      // (top, left) coordinate for the overlay at `pos`.
+      let originPoint = this._getOriginConnectionPoint(originRect, pos);
+      let overlayPoint = this._getOverlayPoint(originPoint, overlayRect, pos);
+      firstOverlayPoint = firstOverlayPoint || overlayPoint;
+
+      // If the overlay in the calculated position fits on-screen, put it there and we're done.
+      if (this._willOverlayFitWithinViewport(overlayPoint, overlayRect, viewportRect)) {
+        this._setElementPosition(element, overlayPoint);
+        return Promise.resolve();
+      }
+    }
+
+    // TODO(jelbourn): fallback behavior for when none of the preferred positions fit on-screen.
+    // For now, just stick it in the first position and let it go off-screen.
+    this._setElementPosition(element, firstOverlayPoint);
+    return Promise.resolve();
+  }
+
+
+  /** Adds a preferred position to the end of the ordered preferred position list. */
+  addPreferredPosition(pos: RelativePosition): void {
+    this._preferredPositions.push(pos);
+  }
+
+
+  /**
+   * Gets the horizontal (x) "start" dimension based on whether the overlay is in an RTL context.
+   * @param rect
+   */
+  _getStartX(rect: ClientRect): number {
+    return this._isRtl ? rect.right : rect.left;
+  }
+
+  /**
+   * Gets the horizontal (x) "end" dimension based on whether the overlay is in an RTL context.
+   * @param rect
+   */
+  _getEndX(rect: ClientRect): number {
+    return this._isRtl ? rect.left : rect.right;
+  }
+
+
+  /**
+   * Gets the (x, y) coordinate of a connection point on the origin based on a relative position.
+   * @param originRect
+   * @param pos
+   */
+  private _getOriginConnectionPoint(originRect: ClientRect, pos: RelativePosition): Point {
+    const originStartX = this._getStartX(originRect);
+    const originEndX = this._getEndX(originRect);
+
+    let x: number;
+    if (pos.originX == 'center') {
+      x = originStartX + (originRect.width / 2);
+    } else {
+      x = pos.originX == 'start' ? originStartX : originEndX;
+    }
```

Comment:
Nitpick: You could split this into `_getOriginConnectionX` and `_getOriginConnectionY` methods to make it a bit more focused.